pkg/agent/client.go

```diff
@@ -399,21 +393,38 @@ func (a *Client) Serve() {
 					if err := a.Send(resp); err != nil {
 						klog.ErrorS(err, "close response failure")
 					}
+				}
 
-					close(dataCh)
-					a.connManager.Delete(connID)
-				},
+				close(dataCh)
+				a.connManager.Delete(connID)
 			}
 			a.connManager.Add(connID, ctx)
 
-			resp.GetDialResponse().ConnectID = connID
-			if err := a.Send(resp); err != nil {
-				klog.ErrorS(err, "stream send failure")
-				continue
-			}
+			go func() {
+				start := time.Now()
+				conn, err := net.Dial(dialReq.Protocol, dialReq.Address)
+				if err != nil {
+					a.connManager.Delete(connID)
+					resp.GetDialResponse().Error = err.Error()
+					if err := a.Send(resp); err != nil {
+						klog.ErrorS(err, "could not send stream")
+					}
+					return
+				}
+				metrics.Metrics.ObserveDialLatency(time.Since(start))
+
+				ctx.conn = conn
+				close(dialDone)
```

Comment:
Let's call this in a defer func in the beginning of the function, otherwise we might miss the `close` on early returns (e.g., the `return` at line 412), and that will lead to leaks.

Reply:
Moved the close as a deferred call directly to the beginning of the function as requested.